File: algorithms/code/leetcode/lc378_kth_smallest_element_in_a_sorted_matrix/LC378KthSmallestElementInASortedMatrix.java
package leetcode.lc378_kth_smallest_element_in_a_sorted_matrix;

import java.util.Collections;
import java.util.PriorityQueue;

/**
 * https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
 * #medium
 */
public final class LC378KthSmallestElementInASortedMatrix {
    public int kthSmallest(final int[][] matrix, final int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>(k, Collections.reverseOrder());
        for (int[] row : matrix) {
            for (int cell : row) {
                heap.add(cell);
                if (heap.size() > k) {
                    heap.poll();
                }
            }
        }
        return heap.poll();
    }
}
package leetcode.lc378_kth_smallest_element_in_a_sorted_matrix;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class LC378KthSmallestElementInASortedMatrixTests {
    @Test
    public void testExample() throws Exception {
        int[][] matrix = { { 1, 5, 9 }, { 10, 11, 13 }, { 12, 13, 15 } };
        assertEquals(13, new LC378KthSmallestElementInASortedMatrix().kthSmallest(matrix, 8));
    }
}

File: algorithms/code/leetcode/lc378_kth_smallest_element_in_a_sorted_matrix/lc378_kth_smallest_element_in_a_sorted_matrix.rs
// https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
// #medium

use std::collections::BinaryHeap;

pub struct Solution;

impl Solution {
    pub fn kth_smallest(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
        let mut heap: BinaryHeap<i32> = BinaryHeap::new();
        for row in matrix.iter() {
            for cell in row.iter() {
                heap.push(*cell);
                if heap.len() as i32 > k {
                    heap.pop();
                }
            }
        }
        heap.pop().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let matrix = vec![vec![1, 5, 9], vec![10, 11, 13], vec![12, 13, 15]];
        assert_eq!(13, Solution::kth_smallest(matrix, 8));
    }
}
